### PR TITLE
Generate deterministic IDs for CompleteEntities

### PR DESCRIPTION
**Pull request by lucaspcram**
### Description:

Added a generator to compute a deterministic ID for a `CompleteEntity`. This ID is based on the entity properties.

### Potential Impact:

N/A

### Unit Test Approach:

Tests added in `EntityIdentifierGeneratorTest`.

### Test Results:

Tests pass.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteArea.java [hunk 1]
```diff
@@ -140,6 +141,17 @@ public long getIdentifier()
         return this.tags;
     }
 
+    @Override
+    public String getWKT()
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteArea.java]: You could call this one `toWkt()` and then use `polygon.toWkt()`?

File: src/main/java/org/openstreetmap/atlas/utilities/identifiers/EntityIdentifierGenerator.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+     * @return the hash
+     */
+    public long generate(final CompleteEntity<?> entity)
+    {
+        final String entityString = getPropertyString(entity)
+                + getTypeSpecificPropertyString(entity);
+        UUID entityHash = UUID.nameUUIDFromBytes(entityString.getBytes());
+
+        long shortHash = entityHash.getMostSignificantBits();
+        while (!isHashSafeToUse(shortHash))
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/utilities/identifiers/EntityIdentifierGenerator.java]: Want to add a safeguard here, to protect the unlikely event of a (close to?) infinite loop?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteRelation.java [hunk 3]
```diff
@@ -288,6 +289,17 @@ public String toString()
...
 
+    @Override
+    public String toWkt()
+    {
+        if (this.bounds == null)
+        {
+            return null;
+        }
+        final WktPolygonConverter converter = new WktPolygonConverter();
+        return converter.convert(this.bounds);
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteRelation.java]: Use `toWkt` here too?

File: src/main/java/org/openstreetmap/atlas/geography/PolyLine.java [hunk 4]
```diff
@@ -52,19 +52,21 @@
...
-    protected static final int SIMPLE_STRING_LENGTH = 200;
-    private static final Logger logger = LoggerFactory.getLogger(PolyLine.class);
-    private static final String IMMUTABLE_POLYLINE = "A polyline is immutable";
-
     public static final PolyLine TEST_POLYLINE = new PolyLine(Location.TEST_3, Location.TEST_7,
             Location.TEST_4, Location.TEST_1, Location.TEST_5);
     public static final PolyLine TEST_POLYLINE_2 = new PolyLine(Location.TEST_1, Location.TEST_5,
             Location.TEST_4, Location.TEST_3, Location.TEST_7);
     public static final PolyLine CENTER = new PolyLine(Location.CENTER);
+    public static final PolyLine SIMPLE_POLYLINE = new PolyLine(Location.forString("1,1"),
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/PolyLine.java]: Sonar wants this to be protected... I think we can live with it being public.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteRelation.java (hunk 3 above)
Comment (lucaspcram) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/complete/CompleteRelation.java]: Whoops missed that.

**Pull request merged**